WalletWasabi/Wallets/FilterProcessor/BlockDownloadService.cs

```
@@ -0,0 +1,370 @@
+			lock (Lock)
+			{
+				while (BlocksToDownload.TryDequeue(out Request? request, out _))
+				{
+					_ = request.Tcs.TrySetResult(CanceledResult.Instance);
+				}
+			}
+		}
+	}
+
+	private async Task HandleSingleBlockTaskAsync(Request request, CancellationToken cancellationToken)
+	{
+		RequestResponse response = await DownloadSingleBlockAsync(request, cancellationToken).ConfigureAwait(false);
+
+		if (response.Result is SuccessResult)
+		{
+			_ = request.Tcs.TrySetResult(response.Result);
+		}
+		else
+		{
+			Logger.LogDebug($"Attempt to download block {request.BlockHash} (height: {request.Priority.BlockHeight}) failed.");
+
+			// The block might have been removed concurrently if a reorg occurred.
+			if (!request.Tcs.TrySetResult(response.Result))
+			{
+				IResult opResult = await request.Tcs.Task.ConfigureAwait(false);
+				Logger.LogDebug($"Failed to set result for block '{request.BlockHash}' (height: {request.Priority.BlockHeight}). Result is already: {opResult}");
+			}
+		}
+	}
+
+	/// <summary>
+	/// Downloads a single block from a block source, or gets the block from the file-system cache if available.
+	/// </summary>
+	private async Task<RequestResponse> DownloadSingleBlockAsync(Request request, CancellationToken cancellationToken)
+	{
+		Logger.LogTrace($"Trying to download {request.BlockHash} (height: {request.Priority.BlockHeight}).");
+
+		try
+		{
+			// Try get the block from the file-system storage.
+			Block? block = await FileSystemBlockRepository.TryGetAsync(request.BlockHash, cancellationToken).ConfigureAwait(false);
+			if (block is not null)
+			{
+				return new RequestResponse(request, new SuccessResult(block, EmptySourceData.FileSystemCache));
+			}
+
+			SuccessResult? successResult = null;
+			ISourceData? failureSourceData = null;
+
+			if (request.SourceRequest is FullNodeSourceRequest)
+			{
+				if (TrustedFullNodeBlockProviders.Length == 0)
+				{
+					return new RequestResponse(request, NoSuchProviderResult.Instance);
```

Interesting. The plan is to introduce some kind of new behavior? What happens if I set my local node as source and it cannot provides? We are not falling back to random nodes anymore?